### PR TITLE
Migrate standalone deployment to workload identity on GCP

### PR DESCRIPTION
**Pull request by Bobgy**
Part of https://github.com/kubeflow/pipelines/issues/1691

Current status:
* [x] Migration works well with our entire test suite
* [ ] Needs documentation updates afterwards for users
    * [ ] Documentation in https://www.kubeflow.org/docs/pipelines/standalone-deployment-gcp/
    * [ ] README.md updates for all references to gcp authentication (use_gcp_secret)
    * [ ] We likely need another doc to recommend which authentication they should use with KFP

Changes:
* Added a bash script for conveniently binding workload identity
* Changed test infra to use workload identity
* Changed samples to not use `use_gcp_secret`
* Changed container builder to not use gcp secret (Future work of supporting both is tracked in https://github.com/kubeflow/pipelines/issues/2702)
* Changed tensorboard template to not use gcp secret

Note, I didn't completely remove test infra for other than workload identity, because I think it still makes sense to also test KFP in for example full scope cluster. So that we can easily add another test configuration in prow with different ENV to use a different one.

/assign @IronPan 
/cc @gaoning777 
/cc @rmgogogo

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/pipelines/2619)

### CODE REVIEW

File: manifests/kustomize/gcp-workload-identity-setup.sh [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+# limitations under the License.
+
+set -e
+
+# Google service Account (GSA)
+SYSTEM_GSA=${SYSTEM_GSA:-$CLUSTER_NAME-kfp-system}
+USER_GSA=${USER_GSA:-$CLUSTER_NAME-kfp-user}
+
+# Kubernetes Service Account (KSA)
+SYSTEM_KSA=(ml-pipeline ml-pipeline-ui)
```
Comment (Bobgy) [inline manifests/kustomize/gcp-workload-identity-setup.sh]: Does any other KSA need to be bound to kfp system service account?

**Comment by gaoning777** [comment]
Is this PR ready to be reviewed?

**Comment by Bobgy** [comment]
> Is this PR ready to be reviewed?
Yes, but only review the script used to set up workload identity.

I ran samples successfully with it. However, still struggling to migrate test infra on workload identity.

**Comment by Bobgy** [comment]
It turns out that tests failed before because our image builder docker image was built a year ago with google/cloud-sdk. The old version of cloud sdk doesn't work well with workload identity.

**Comment by Bobgy** [comment]
UPDATE, I removed all usages of use_gcp_secret to pass sample tests.
However, most of the samples need documentation update about how to authorize on GCP

**Comment by gaoning777** [comment]
QQ: for on-prem scenarios, how does the container builder get the credentials?

**Comment by Bobgy** [comment]
@gaoning777 it's not supported in this PR. I think we need to expose more flexible configuration to support it.

Is this container builder used by python components from notebook?

**Comment by gaoning777** [comment]
> @gaoning777 it's not supported in this PR. I think we need to expose more flexible configuration to support it.
> 
> Is this container builder used by python components from notebook?

Yes, currently the container builder function is being used from the notebook with SDK. We probably need to expose more configurations from the SDK.

**Comment by Bobgy** [comment]
Created issue https://github.com/kubeflow/pipelines/issues/2702 to track container builder support

File: manifests/kustomize/gcp-workload-identity-setup.sh (hunk 1 above)
Comment (IronPan) [reply manifests/kustomize/gcp-workload-identity-setup.sh]: Do you know why ml-pipeline need gcp access?

Ideally non of these services would require GCP access, for the purpose of secured multi-user architecture. 
/cc @gaoning777

File: samples/core/exit_handler/exit_handler.py [hunk 2]
```diff
@@ -21,7 +21,7 @@
 def gcs_download_op(url):
     return dsl.ContainerOp(
         name='GCS - Download',
-        image='google/cloud-sdk:216.0.0',
+        image='google/cloud-sdk:latest',
```
Comment (IronPan) [inline samples/core/exit_handler/exit_handler.py]: Is this necessary? If it need a later version, can we ping to that version instead of using latest?

Comment (Bobgy) [reply samples/core/exit_handler/exit_handler.py]: I can verify again, that was my guess, but it seemed we may not need this change.

File: manifests/kustomize/gcp-workload-identity-setup.sh (hunk 1 above)
Comment (Bobgy) [reply manifests/kustomize/gcp-workload-identity-setup.sh]: I don't know. Let me try removing this.

File: test/deploy-pipeline-lite.sh [hunk 3]
```diff
@@ -49,8 +51,28 @@ kustomize edit set image gcr.io/ml-pipeline/inverse-proxy-agent=${GCR_IMAGE_BASE
...
+
+  gcloud projects add-iam-policy-binding $PROJECT \
+    --member="serviceAccount:$SYSTEM_GSA@$PROJECT.iam.gserviceaccount.com" \
+    --role="roles/editor"
+  gcloud projects add-iam-policy-binding $PROJECT \
+    --member="serviceAccount:$USER_GSA@$PROJECT.iam.gserviceaccount.com" \
+    --role="roles/editor"
+
+  source "$DIR/../manifests/kustomize/wi-utils.sh"
+  verify_workload_identity_binding "pipeline-runner" $NAMESPACE
```
Comment (Bobgy) [inline test/deploy-pipeline-lite.sh]: Note, after a lot of rounds of trial and error, I figured out the following:
1. workload identity binding doesn't work for the first few requests no matter how long you wait for it
2. after you try to use the binding a few times, it starts to work stably

That's why I call these verification functions for each service account that's important in the test suite.

**Comment by Bobgy** [comment]
Note, current test failure might be a bug in test infra: image is cached based on PR commit hash, but it should be based on merge commit hash.

**Comment by Bobgy** [comment]
Got the following error in sample test (volume_snapshot_ops)
```
run-sample-tests-loop(12:volume_snapshot_ops):	google.auth.exceptions.TransportError: ("Failed to retrieve http://metadata.google.internal/computeMetadata/v1/instance/service-accounts/default/?recursive=true from the Google Compute Enginemetadata service. Status: 500 Response:\nb'GKE Metadata Server encountered an error: Could not recursively fetch uri\\n'", <google.auth.transport.requests._Response object at 0x7f7508e246d8>)
```

I'm not sure this is related, will pin google-cloud/sdk image first

**Comment by Bobgy** [comment]
Verfied tests are flaky with google/cloud-sdk:216.0.0

**Comment by Bobgy** [comment]
Verify testing again for flakiness
/test kubeflow-pipeline-e2e-test
/test kubeflow-pipeline-sample-test

File: manifests/kustomize/gcp-workload-identity-setup.sh (hunk 1 above)
Comment (Bobgy) [reply manifests/kustomize/gcp-workload-identity-setup.sh]: Removed the binding and tests keep passing. I think this is good.

File: samples/core/exit_handler/exit_handler.py (hunk 2 above)
Comment (Bobgy) [reply samples/core/exit_handler/exit_handler.py]: I pinned to latest version and verified this is required. Using old cloud-sdk makes tests flaky.

File: test/deploy-pipeline-lite.sh (hunk 3 above)
Comment (IronPan) [reply test/deploy-pipeline-lite.sh]: That's interesting. Should we surface this to GKE team?

Comment (Bobgy) [reply test/deploy-pipeline-lite.sh]: Yes, I think so. I will send an email next week

**Comment by Bobgy** [comment]
/test kubeflow-pipeline-sample-test
sidecar sample is flaky

**Comment by Bobgy** [comment]
container builder seems to be flaky at fetching some image: https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/kubeflow_pipelines/2619/kubeflow-pipeline-sample-test/1206752832519147523#1:build-log.txt%3A4165
secret sample has the following error, I haven't yet figured out if that's a workload identity transient error or not
https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/kubeflow_pipelines/2619/kubeflow-pipeline-sample-test/1206752832519147523#1:build-log.txt%3A4000

Test again to verify
/test kubeflow-pipeline-sample-test

**Comment by Bobgy** [comment]
Verify again for flakiness
/test kubeflow-pipeline-sample-test

File: manifests/kustomize/wi-utils.sh [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+# Bind KSA to GSA through workload identity.
+# Documentation: https://cloud.google.com/kubernetes-engine/docs/how-to/workload-identity
+function bind_gsa_and_ksa {
+  local gsa=${1}
+  local ksa=${2}
+  local project=${3:-$PROJECT_ID}
+  local gsa_full="$gsa@$project.iam.gserviceaccount.com"
+  local namespace=${4:-$NAMESPACE}
+
+  gcloud iam service-accounts add-iam-policy-binding $gsa_full \
```
Comment (Ark-kun) [inline manifests/kustomize/wi-utils.sh]: Is it possible to add retries here? This command is a source of some test flakiness:

```
++ gcloud iam service-accounts add-iam-policy-binding test-argo@ml-pipeline-test.iam.gserviceaccount.com '--member=serviceAccount:ml-pipeline-test.svc.id.goog[kubeflow/test-runner]' --role=roles/iam.workloadIdentityUser
ERROR: (gcloud.iam.service-accounts.add-iam-policy-binding) ABORTED: There were concurrent policy changes. Please retry the whole read-modify-write with exponential backoff.
```

Comment (Bobgy) [reply manifests/kustomize/wi-utils.sh]: Sure, let me put up a PR

Comment (Bobgy) [reply manifests/kustomize/wi-utils.sh]: https://github.com/kubeflow/pipelines/pull/3008

File: sdk/python/kfp/containers/_k8s_job_helper.py [hunk 5]
```diff
@@ -48,24 +48,10 @@ def _create_k8s_job(self, yaml_spec):
...
-                                       args = yaml_spec['spec']['containers'][0]['args'],
-                                       volume_mounts = [k8s_client.V1VolumeMount(
-                                           name=yaml_spec['spec']['containers'][0]['volumeMounts'][0]['name'],
-                                           mount_path=yaml_spec['spec']['containers'][0]['volumeMounts'][0]['mountPath'],
-                                       )],
-                                       env = [k8s_client.V1EnvVar(
-                                           name=yaml_spec['spec']['containers'][0]['env'][0]['name'],
-                                           value=yaml_spec['spec']['containers'][0]['env'][0]['value'],
-                                       )])
+                                       args = yaml_spec['spec']['containers'][0]['args'])
```
Comment (Ark-kun) [inline sdk/python/kfp/containers/_k8s_job_helper.py]: I'm a bit concerned about this change to the container-building functions for non-GKE deployments.
For container-building functions there might be no way for the user to supply secrets.

Comment (Bobgy) [reply sdk/python/kfp/containers/_k8s_job_helper.py]: Yes, I also think it's important to fix. However, I don't currently have bandwidth. Let's get this scheduled soon then.

File: samples/core/secret/secret.py [hunk 6]
```diff
@@ -57,11 +56,8 @@ def use_gcp_api_op():
 def secret_op_pipeline(url='gs://ml-pipeline-playground/shakespeare1.txt'):
   """A pipeline that uses secret to access cloud hosted resouces."""
 
-  gcs_read_task = gcs_read_op(url).apply(
```
Comment (Ark-kun) [inline samples/core/secret/secret.py:60]: The use of secrets seems to be the core feature demonstrated by this sample. Perhaps this should not be removed.

Comment (Bobgy) [reply samples/core/secret/secret.py:60]: Yes, you are right. Current situation is that we are not running tests that use a service account token. Shall we remove this sample from presubmit/postsubmit tests?

**Pull request merged**